[19.0][ADD] partner_identification_kyc

bosd opened the pull request:
Add module to use the partner_identification number model for KYC processes.

#2218
#2224
#2225

bosd committed:
[ADD] partner_identification_kyc

bosd committed:
[ DO NOT MERGE TEST REQUIREMENTS]

bosd committed:
partner_identification_kyc: Support child contacts

bosd committed:
partner_identification_kyc: define header

Commit:
partner_identification_kyc: improve tests

BhaveshHeliconia commented on partner_identification_kyc/data/activity_type_data.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
Add `noupdate="1"`

BhaveshHeliconia commented on partner_identification_kyc/demo/identification_number_demo.xml:
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
Add `noupdate="1"`

BhaveshHeliconia commented on partner_identification_kyc/demo/identification_number_demo.xml:
```diff
@@ -0,0 +1,103 @@
+        <field name="category_id" ref="kyc_identification_category" />
+        <field name="name">KYC-001-DRAFT</field>
+        <field name="status">draft</field>
+        <field name="validity_start" eval="context_today().strftime('%Y-%m-01')" />
```
Replace `validity_start`, `issuer_id` and `validity_end` fields with correct ones

BhaveshHeliconia commented on partner_identification_kyc/models/res_partner.py:
```diff
@@ -0,0 +1,164 @@
+            {
+                "partner_id": partner.id,
+                "category_id": kyc_category.id,
+                "name": f"KYC-{partner.id}-{sequence_code}",
```
`KYC` is being duplicated in `ID record` name

BhaveshHeliconia commented on partner_identification_kyc/views/identification_number_views.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="arch" type="xml">
+            <search>
+                <filter
+                    name="kyc_category_filter"
```
Missing in view(view inheritance is missing).

BhaveshHeliconia commented on partner_identification_kyc/models/res_partner_id_category.py:
```diff
@@ -0,0 +1,12 @@
+        default=False,
+        help="If checked, KYC checks can be performed on child contacts of a company.",
+    )
+    responsible_id = fields.Many2one("res.users", string="Responsible")
```
Currently not being used. We can use `responsible_user_id` from `responsible_user_id` module if needed in future.

BhaveshHeliconia commented on partner_identification_kyc/models/res_partner.py:
```diff
@@ -0,0 +1,164 @@
+            ],
+        }
+
+    def _create_kyc_record(self, partner):
```
Already defined in `ResPartner`. We should omit `partner` argument and use `self` directly.

BhaveshHeliconia commented on partner_identification_kyc/views/res_partner_id_category_view.xml:
```diff
@@ -0,0 +1,16 @@
+        />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='code']" position="after">
+                <field name="enable_on_child_contacts" invisible="code!='KYC'" />
```
Condition looks hard-coded. Need possible workaround here.

Commit:
[ DO NOT MERGE TEST REQUIREMENTS]

Commit:
[ADD] partner_identification_kyc

Add a module providing Know Your Customer (KYC) identification for
partners, built on top of partner_identification:

- a dedicated "KYC" identification category with its check activity type
- a "Request KYC" button on the partner form and a helper to ensure a
  KYC record exists (e.g. for API-created partners)
- computed KYC validity / button-visibility, optionally enabled on child
  contacts
- a KYC filter on the id_numbers views and demo data

BhaveshHeliconia reviewed (commented):
`enable_on_child_contacts` field is currently only being used with KYC category, so we may use one boolean field to distinguish KYC category from others and also restore `invisible="code!='KYC'"` with corrected condition expression.

`ID Number` field is not setting with  automatically generated sequence while trying to create KYC `ID records` manually using one2many field.

Commit:
[IMP] partner_identification_kyc: address review feedback

- Add an is_kyc boolean on res.partner.id_category to mark the KYC category
  instead of hard-coding the 'KYC' code; enable_on_child_contacts is now only
  shown for KYC categories (invisible="not is_kyc"), and the search filter
  matches on is_kyc.
- Auto-assign the identification number from the kyc.identification sequence in
  res.partner.id_number.create, so KYC records created manually via the
  one2many no longer require a hand-typed ID Number (name is relaxed for KYC in
  the form view).
- Drop the duplicated 'KYC-' prefix from the generated ID Number (the category
  already conveys KYC).
- Tests: cover manual auto-sequence and provided-name paths.

bosd commented:
@BhaveshHeliconia thanks for the thorough review! Addressed in a26692c5b. Point by point:

**Design points**
- *`enable_on_child_contacts` only relevant for KYC* — added an `is_kyc` boolean on `res.partner.id_category` to distinguish the KYC category, and `enable_on_child_contacts` is now `invisible="not is_kyc"` (no hard-coded `code == 'KYC'`). The KYC search filter now matches on `is_kyc` too.
- *ID Number not auto-set from the sequence on manual creation* — `res.partner.id_number.create` now auto-assigns the `kyc.identification` sequence for KYC records when no name is given, and the form relaxes the `name` requirement for KYC. Works for both the *Request KYC* button and manual one2many creation.
- *`KYC` duplicated in the ID record name* — dropped the redundant `KYC-` prefix; the number now comes purely from the sequence (the category already conveys KYC).

**Already fixed in the latest push** (the review predates a force-push, so these showed as outdated):
- `noupdate="1"` is present on both `data/activity_type_data.xml` and `demo/identification_number_demo.xml`.
- The demo uses the correct base fields (`valid_from` / `valid_until` / `partner_issued_id`).
- `responsible_user_id` is no longer present.
- No method takes a redundant `partner` argument (all use `self`).
- The identification-number search view inherits `view_partner_id_numbers_search` correctly.

Added tests for the manual auto-sequence and the provided-name paths. Could you re-review when you have a moment? 🙏

Commit:
[FIX] partner_identification_kyc: update KYC name assertion in tests

test_action_request_kyc_with_custom_sequence still asserted the old
"KYC-" prefixed ID Number; the number now comes purely from the
sequence. Assert the name is set and no longer duplicates the category.

BhaveshHeliconia reviewed (commented):
- Keep prefix back in `seq_kyc_identification` or modify `create` method logic.
- `ID Number` field is required in form view, so unable to save form without it.

Commit:
[IMP] partner_identification_kyc: prefix KYC sequence and prefill ID Number

Addresses the second review round:
- Configure the "KYC" prefix on seq_kyc_identification (instead of
  hard-coding it in create), so generated ID Numbers read e.g. KYC001.
- Prefill the (required) ID Number from the sequence via an onchange when
  the KYC category is selected, so a KYC record can be created manually
  through the one2many without typing the number. create() keeps assigning
  it for the Request KYC button and programmatic creation.
- Drop the now-unnecessary form-view required modifier and the unused
  category_is_kyc related field.
- Tests: assert the KYC prefix and add an onchange/Form test.

bosd commented:
@BhaveshHeliconia thanks for testing! Both points addressed in 349dcd16d:

- **Prefix** — restored the `KYC` prefix, but on the sequence itself (`seq_kyc_identification` now has `<field name="prefix">KYC</field>`) rather than hard-coded in `create`. Generated ID Numbers now read e.g. `KYC001`.
- **Required ID Number blocking manual save** — instead of trying to relax the (model-level) required field in the view, the ID Number is now *prefilled* from the sequence via an `onchange` the moment the KYC category is selected, so the required field is satisfied and you don't have to type it. `create()` still assigns it for the *Request KYC* button and programmatic/import creation. I dropped the earlier form-view `required` modifier and the unused `category_is_kyc` field since they're no longer needed.

Added a `Form`-based test covering the onchange prefill. Could you re-test when you have a moment? 🙏

Commit:
[FIX] partner_identification_kyc: test onchange directly instead of Form

common.Form construction errored in the Odoo 19 CI; call the onchange
method on a new() record instead, which tests the same prefill behaviour.

BhaveshHeliconia reviewed (approved):
LGTM!
